extensions/amp-story/1.0/media-performance-metrics-service.js

```diff
@@ -0,0 +1,413 @@
+    const mediaEntry = this.getNewMediaEntry_(media, unlisteners);
+    this.setMediaEntry_(media, mediaEntry);
+
+    // Checks if the media already errored (eg: could have failed the source
+    // selection).
+    if (
+      media.error ||
+      media[MEDIA_LOAD_FAILURE_SRC_PROPERTY] === media.currentSrc
+    ) {
+      mediaEntry.metrics.error = media.error ? media.error.code : 0;
+      mediaEntry.status = Status.ERRORED;
+    }
+  }
+
+  /**
+   * Stops recording, computes, and sends performance metrics collected for the
+   * given media element.
+   * @param {!HTMLMediaElement} media
+   */
+  stopMeasuring(media) {
+    const mediaEntry = this.getMediaEntry_(media);
+
+    if (!mediaEntry) {
+      return;
+    }
+
+    mediaEntry.unlisteners.forEach(unlisten => unlisten());
+    this.deleteMediaEntry_(media);
+
+    switch (mediaEntry.status) {
+      case Status.PLAYING:
+        this.addWatchTime_(mediaEntry);
+        break;
+      case Status.WAITING:
+        this.addRebuffer_(mediaEntry);
+        break;
+    }
+
+    this.sendMetrics_(mediaEntry);
+  }
+
+  /**
+   * @param {!MediaEntryDef} mediaEntry
+   * @private
+   */
+  sendMetrics_(mediaEntry) {
+    const {metrics} = mediaEntry;
+
+    // If the media errored.
+    if (metrics.error !== null) {
+      this.performanceService_.tickDelta('smerr', metrics.error || 0);
+      this.performanceService_.flush();
+      return;
+    }
+
+    // If the user was on the video for less than one second, ignore the metrics
+    // (eg: users tapping through a story, or scrolling through content).
+    if (
+      !metrics.jointLatency &&
+      Date.now() - mediaEntry.timeStamps.start < MINIMUM_TIME_THRESHOLD_MS
+    ) {
+      return;
+    }
+
+    // If the playback did not start.
+    if (!metrics.jointLatency) {
+      this.performanceService_.tickDelta('smerr', 5 /* Custom error code */);
+      this.performanceService_.flush();
+      return;
+    }
+
+    const rebufferRate =
+      Math.round(
+        (metrics.rebufferTime / (metrics.rebufferTime + metrics.watchTime)) *
+          1000
+      ) / 1000;
+
+    this.performanceService_.tickDelta('smjl', metrics.jointLatency);
+    this.performanceService_.tickDelta('smwt', metrics.watchTime);
+    this.performanceService_.tickDelta('smrb', metrics.rebuffers);
+    this.performanceService_.tickDelta('smrbr', rebufferRate);
+    if (metrics.rebuffers) {
+      this.performanceService_.tickDelta(
+        'smmtbrb',
+        Math.round(metrics.watchTime / metrics.rebuffers)
+      );
+    }
+    this.performanceService_.flush();
+  }
+
+  /**
+   * @param {!HTMLMediaElement} media
+   * @return {!MediaEntryDef}
+   * @private
+   */
+  getMediaEntry_(media) {
+    return this.mediaMap_[media[ID_PROPERTY]];
+  }
+
+  /**
+   * @param {!HTMLMediaElement} media
+   * @param {!MediaEntryDef} mediaEntry
+   * @private
+   */
+  setMediaEntry_(media, mediaEntry) {
+    media[ID_PROPERTY] = media[ID_PROPERTY] || this.mediaId_++;
```

Comment:
That would work nicely with the media pool because we guarantee our IDs to be unique, but the issue is that the following example would be valid AMP:

```
  <amp-video width="100" height="20" id="foo">
  </amp-video>
  <amp-video width="100" height="20" id="foo">
  </amp-video>
```

Reply:
It's valid AMP but is not actually valid HTML.  It goes against the HTML spec:

> The id attribute specifies its element's unique identifier (ID). The value must be unique amongst all the IDs in the element's home subtree and must contain at least one character.
[source](https://www.w3.org/TR/2011/WD-html5-20110525/elements.html#the-id-attribute)

...and it also fails in validator.w3.org.

Perhaps more importantly, it just means we would drop metrics for one of those two videos, which on a large scale, I'm not sure how important that is.  Open to not doing this, but it honestly feels to me like it should be fine.

Reply:
It's your call of course, but the thing I'm not sure about is what's the benefit of relying on IDs, since we'd still need our `mediaId` for videos that don't have it.
The way I understand it is that we'd be adding one line of code just to open a way for potential ID collisions :( Am I missing something?

Reply:
Ah, I was thinking of it from the perspective of the memory overhead of allocating yet another property to reflect the same thing.  Of course it's not very much, I just wonder how many collisions we'd really run into also